### DISCUSSION TITLE
Nodejs install brings packages for all environments - 66MB size

### DISCUSSION THREAD

**Discussion by adityapatadia** [envkey/envkey#18] [Ideas and feedback]
When envkey is installed, it brings packages for all environments. Ideally it should bring the binary for given OS on which it's installed. Currently envkey package is biggest size in `node_modules` folder becasue of this
<img width="444" alt="Screenshot 2022-05-14 at 3 06 19 PM" src="https://user-images.githubusercontent.com/1086617/168420191-737d2ea2-6f76-47f7-b35e-a76a7aa06085.png">
.

**Comment by danenania** [comment]
Hey @adityapatadia, this is a deliberate tradeoff to make the SDKs work seamlessly cross-platform. While there may be some ways around this, I don't think any will be as simple and reliable as the current approach.

An alternative is using [envkey-source](https://docs-v2.envkey.com/docs/envkey-source) directly to avoid the extra weight.

**Comment by adityapatadia** [comment]
My idea was not to reduce the number of builds. Cross platform is super important. I know it would be little more work but downloding required binary at the time of installation like https://github.com/lovell/sharp would work best. Please consider in future releases.

If possible, please also provide instructions on how to use envkey-source inside node package.

**Comment by danenania** [comment]
> downloding required binary at the time of installation

I've considered it, but there are some potential edge cases with downloading at the time of installation, including use of bundlers and locked down networks, that have made me hesitant to go this route.

> If possible, please also provide instructions on how to use envkey-source inside node package.

To use envkey-source directly, you wouldn't need to call it from inside the package. You could just prefix your app start command like this:

```
$ envkey-source -- npm start
```

That would give you the same behavior as the node package. You could also add auto-reloads when a variable changes:

```
$ es -w -- npm start
```

(Using `es` alias here which is equivalent to `envkey-source`)

In development you can also tack on `-c` for local caching so you can keep working with a lost internet connection:

```
$ es -wc -- npm start
```

**Reply by adityapatadia** [reply]
How does signals like SIGTERM and SIGINT passing work to nodejs? We tried using Doppler CLI in past and in a kubernetes cluster, those signals were not passed correctly.

**Reply by danenania** [reply]
SIGTERM and SIGINT are caught and then passed on to the watched process. The process is then given some time to handle the signal and shut down (5 seconds by default, but configurable) before it is killed. 

Another option if you don't want to deal with process watching is to set your EnvKey vars as environment variables in the current shell before starting your app. That would look something like:

```
eval $(envkey-source)
npm start
```

**Comment by adityapatadia** [comment]
How will this work with kubernetes? Any idea?

**Reply by danenania** [reply]
It fits very smoothly into Kubernetes and there isn't a whole lot to it: https://docs-v2.envkey.com/docs/kubernetes

Lmk if anything is unclear there.

**Reply by adityapatadia** [reply]
There is some issue with Kubernetes and when we use `es -- node server.js` it just does not pass signals to nodejs process. We faced same issues with Doppler CLI.

**Reply by danenania** [reply]
This sounds like it could be related to the issue discussed here: https://github.com/envkey/envkey/issues/3

If that's the case, the conclusion reached there was to use one of these libraries to ensure proper signal handling:

- https://github.com/Yelp/dumb-init
- https://github.com/krallin/tini

**Reply by danenania** [reply]
Or you could also use the `eval $(envkey-source)` approach I described above.

**Reply by danenania** [reply]
Hey @adityapatadia were you able to get this working?

**Reply by adityapatadia** [reply]
We think the `eval` route should work fine but due to changes required in our environment and testing, we are not going ahead with this for now. We have a goal to reduce docker image size and this nodejs package is definitely one of the largest. If this becomes big problem, we will try and move to this method.

Thanks a ton for your help so far :)

**Comment by adityapatadia** [comment]
Hey,
I am re-visiting this. I have an idea, in our dockerfile, post installation steps, can we delete the files which are there for other architectures? We know our production build is for amd64 linux and we don't need other binaries.

**Reply by adityapatadia** [reply]
I ended up doing this in my Dockerfile: `cd ./node_modules/envkey/ext && find . ! -name "*_linux_amd64" ! -path . -type d -exec rm -rf {} + `

**Reply by danenania** [reply]
Cool, that works. 👍  There is no requirement for binaries to be present except for the current platform, so there's no problem with deleting the unneeded ones.
